ostd/src/logger.rs

```diff
@@ -10,96 +10,74 @@
 //!
 //! IRQs are disabled while printing. So do not print long log messages.
 
+use core::str::FromStr;
+
 use log::{LevelFilter, Metadata, Record};
+use spin::Once;
+
+static LOGGER: Logger = Logger::new();
 
-use crate::{
-    boot::{kcmdline::ModuleArg, kernel_cmdline},
-    timer::Jiffies,
-};
+struct Logger {
+    backend: Once<&'static dyn log::Log>,
+}
 
-const LOGGER: Logger = Logger {};
+impl Logger {
+    const fn new() -> Self {
+        Self {
+            backend: Once::new(),
+        }
+    }
+}
 
-struct Logger {}
+/// Sets the backend for the global logger, allowing upper-level users to register
+/// their own implemented loggers.
+///
+/// This method only allows setting once; subsequent attempts to set it again will have no effect.
+pub fn set_logger(new_logger: &'static dyn log::Log) {
+    LOGGER.backend.call_once(|| new_logger);
+}
 
 impl log::Log for Logger {
     fn enabled(&self, metadata: &Metadata) -> bool {
+        if let Some(logger) = self.backend.get() {
+            return logger.enabled(metadata);
+        };
+
+        // Default instruction.
         metadata.level() <= log::max_level()
     }
 
     fn log(&self, record: &Record) {
-        if !self.enabled(record.metadata()) {
```

Comment:
This line should not be removed unless `Log::enabled` is an empty function (however, it will be, if you fix the comment above). This is because `Log::enabled` documents [say](https://docs.rs/log/latest/log/trait.Log.html#tymethod.enabled):
> For implementors
>
> This method isn’t called automatically by the log! macros. It’s up to an implementation of the Log trait to call enabled in its own log method implementation to guarantee that filtering is applied.